### PR TITLE
Refactor ExportSampleLogsToHDF5 to use dtype function.

### LINKED ISSUE #22554: Replace isinstance calls in ExportSampleLogsToHDF5 with dtype Property method
The algorithm `ExportSampleLogsToHDF5` (and its unit tests) are riddled with calls to `isinstance` to ensure that `h5py` datasets are given the correct dtype.

There are plans to add a method to `PropertyWithValue` to get the corresponding `numpy` dtype string, and export this to Python (see #22553). Once this is implemented, we should be able to remove the method `_dtype_from_property_type` from `ExportSampleLogsToHDF5` and replace it with a call to the new Property method.

(Gotcha: numpy's `str` dtype doesn't get on well with `h5py` - look [here](http://docs.h5py.org/en/latest/strings.html#how-to-store-text-strings) for details of what to use instead)

Similarly, #22555 will make it easier to construct `TimeSeriesProperty`s from Python, so once that is taken care of we'll be able to clean up the unit tests (possibly a matter for a separate issue).

There's a couple of notes on PR #22357

### PR DESCRIPTION
**Pull request by BhuvanBezawada**
**Description of work.**
Cleaned up code in `ExportSampleLogsToHDF5.py` and the corresponding test file `ExportSampleLogsToHDF5Test.py` by removing many of the calls made to `isinstance` and using the newly implemented `dtype` function ( see issue #22553 ).

Exposed a new `PropertyFactory` class to the Python side which allows for `TimeSeriesProperty` objects to be easily created ( see issue #22555 ) using the `createTimeSeries` function which was also implemented.

This PR depends on issue #22553.

**Report to:** @samueljackson92  

**To test:**
Not much to test, just run the unit tests and check that they all pass.

Run `ctest -C Debug -R ExportSampleLogsToHDF5Test -V` for full information when running the tests.

Fixes #22554 . 

Does this update require release notes?
- [ ] Yes
- [x] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/test/python/plugins/algorithms/ExportSampleLogsToHDF5Test.py [hunk 1]
```diff
@@ -98,21 +126,11 @@ def _create_sample_workspace(self):
...
-        elif isinstance(log_value[0], float):
-            prop = FloatTimeSeriesProperty(log_name)
-        elif isinstance(log_value[0], str):
-            prop = StringTimeSeriesProperty(log_name)
-        elif isinstance(log_value[0], bool):
-            prop = BoolTimeSeriesProperty(log_name)
-        else:
-            raise RuntimeError("Unsupported property type {}".format(type(log_value[0])))
-
+        prop = PropertyFactory.createTimeSeries(log_name, log_value)
```
Comment (samueljackson92) [inline Framework/PythonInterface/test/python/plugins/algorithms/ExportSampleLogsToHDF5Test.py]: 👍 Awesome.

File: Framework/PythonInterface/plugins/algorithms/ExportSampleLogsToHDF5.py [hunk 2]
```diff
@@ -49,7 +49,7 @@ def PyExec(self):
                               if prop.name not in blacklist and not self._ignore_property(prop)]
 
             for log_property in log_properties:
-                property_dtype = self._dtype_from_property_type(log_property)
+                property_dtype = log_property.dtype()
```
Comment (samueljackson92) [inline Framework/PythonInterface/plugins/algorithms/ExportSampleLogsToHDF5.py]: 👍 Awesome

**Review by samueljackson92 — CHANGES_REQUESTED** [review]
Code is looking good, but we've got some build failures that look related to these changes that need to be resolved.

**Comment by BhuvanBezawada** [comment]
@samueljackson92 I think one of the build errors is to do with some of the `Py_Check` functions that I'm using - I think it has been taken out of Python3 and no longer works?

File: Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp [hunk 3]
```diff
@@ -138,6 +150,40 @@ PropertyWithValueFactory::create(const std::string &name,
...
+    const std::string &name, const boost::python::object &defaultValue) {
+
+  // Take in the object and try to convert it into a list
+  const boost::python::list pyList = extract<boost::python::list>(defaultValue);
+
+  // Use a PyObject pointer to determine the type stored in the list
+  auto object = boost::python::object(pyList[0]).ptr();
+
+  // Decide which kind of TimeSeriesProperty to return
+  if (PyBool_Check(object) == 1) {
```
Comment (samueljackson92) [inline Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp]: Don't use `PyBool_Check` use: 
```cpp
boost::python::extract<bool>(object).check()
```

Comment (samueljackson92) [reply Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp]: Also why not add a using statement for boost::python?

**Review by samueljackson92 — CHANGES_REQUESTED** [review]

File: Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp (hunk 3 above)
Comment (BhuvanBezawada) [reply Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp]: I tried extracting with boost first and I think the issue was that the `.check()` function accepted 1 as a Boolean value. So if a list of integers was given e.g. [1,2,3,4,5] then a `BoolTimeSeriesProperty` was returned.

Comment (samueljackson92) [reply Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp]: Ah i see. Instead I guess we want to check that we can convert to a std::vector. I think there might be something in the `Converters` namespace that can do this.

Comment (samueljackson92) [reply Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp]: You could always expect a `boost::python::list` and then check size, then check the type of the first item. Here's an example: https://stackoverflow.com/a/3761526

Comment (BhuvanBezawada) [reply Framework/PythonInterface/mantid/kernel/src/Registry/PropertyWithValueFactory.cpp]: Okay I'll give that a go!

**Comment by BhuvanBezawada** [comment]
@samueljackson92 I'm a bit unsure what the errors for the Ubuntu builds are about/why they occur...

**Comment by samueljackson92** [comment]
It's a doc test failure. i.e. the test code being run from the documentation file ExportSampleLogsToHDF5.rst is failing because the algorithm can't write to HDF5 on line 58 in the algorithm.

You can run the doc tests locally like so: http://developer.mantidproject.org/Standards/DocumentationGuideForDevs.html?highlight=doc%20tests#running-documentation-tests-locally

File: Framework/PythonInterface/inc/MantidPythonInterface/kernel/Converters/ContainerDtype.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  } else if (std::is_same<HeldType, std::int8_t>::value) {
+    return "i";
+  } else if (std::is_same<HeldType, std::int16_t>::value) {
+    return "i";
+  } else if (std::is_same<HeldType, std::int32_t>::value) {
+    return "i";
+  } else if (std::is_same<HeldType, std::int64_t>::value) {
+    return "i";
+  } else if (std::is_same<HeldType, long>::value) {
+    return "i";
```
Comment (samueljackson92) [inline Framework/PythonInterface/inc/MantidPythonInterface/kernel/Converters/ContainerDtype.h]: Perhaps now we can just use `is_integral`?

**Review by samueljackson92 — CHANGES_REQUESTED** [review]
Looking pretty good. A few more comments.

File: Framework/PythonInterface/inc/MantidPythonInterface/kernel/Converters/ContainerDtype.h (hunk 4 above)
Comment (BhuvanBezawada) [reply Framework/PythonInterface/inc/MantidPythonInterface/kernel/Converters/ContainerDtype.h]: Yeah I think that's a good plan

File: Framework/PythonInterface/mantid/kernel/src/Exports/TimeSeriesProperty.cpp [hunk 5]
```diff
@@ -35,6 +36,35 @@ void addPyTimeValue(TimeSeriesProperty<TYPE> &self,
...
+// Call the dtype helper function
+template <typename TYPE> std::string dtype(TimeSeriesProperty<TYPE> &self) {
+  return Mantid::PythonInterface::Converters::dtype(self);
+}
+
+// Check for the special case of a string
+template <> std::string dtype(TimeSeriesProperty<std::string> &self) {
+  // Vector of ints to store the sizes of each of the strings
+  std::vector<size_t> string_sizes;
+
```
[the remark this answers is not in the document]
Comment (BhuvanBezawada) [reply Framework/PythonInterface/mantid/kernel/src/Exports/TimeSeriesProperty.cpp]: Ah sorry - missed that! I'll change it now.

**Review by samueljackson92 — APPROVED** [review]

**Comment by martyngigg** [comment]
@BhuvanBezawada I was trying to solve the merge conflict from clang formatting quickly here but there now seems to be an issue with building after merging in master. Could you take a look?

**Comment by BhuvanBezawada** [comment]
Yeah no problem, I'll have a look just now.

**Pull request merged**